src/abstractdatatable/io.jl

```diff
@@ -216,118 +216,60 @@ function Data.isdone(source::DataTable, row, col)
-Data.streamfrom{T}(source::DataTable, ::Type{Data.Field}, ::Type{T}, row, col) =
-    (@inbounds A = Data.streamfrom(source, Data.Column, T, col); return A[row]::T)
+# Data.streamfrom{T <: AbstractVector}(source::DataTable, ::Type{Data.Batch}, ::Type{T}, col) =
+#     (A = source.columns[col]::T; return A)
+Data.streamfrom{T}(source::DataTable, ::Type{Data.Batch}, ::Type{T}, col) =
+    (A = source.columns[col]::AbstractVector{T}; return A)
+Data.streamfrom{T}(source::DataTable, ::Type{Data.Row}, ::Type{T}, row, col) =
+    (A = source.columns[col]::AbstractVector{T}; return A[row]::T)
 
 # DataTable as a Data.Sink
-allocate{T}(::Type{T}, rows, ref) = Array{T}(rows)
-allocate{T}(::Type{Vector{T}}, rows, ref) = Array{T}(rows)
-
-allocate{T}(::Type{Nullable{T}}, rows, ref) =
-    NullableArray{T, 1}(Array{T}(rows), fill(true, rows), isempty(ref) ? UInt8[] : ref)
-allocate{T}(::Type{NullableVector{T}}, rows, ref) =
-    NullableArray{T, 1}(Array{T}(rows), fill(true, rows), isempty(ref) ? UInt8[] : ref)
-
-allocate{S,R}(::Type{CategoricalArrays.CategoricalValue{S,R}}, rows, ref) =
-    CategoricalArray{S,1,R}(rows)
-allocate{S,R}(::Type{CategoricalVector{S,R}}, rows, ref) =
-    CategoricalArray{S,1,R}(rows)
-
-allocate{S,R}(::Type{Nullable{CategoricalArrays.CategoricalValue{S,R}}}, rows, ref) =
-    NullableCategoricalArray{S,1,R}(rows)
-allocate{S,R}(::Type{NullableCategoricalVector{S,R}}, rows, ref) =
-    NullableCategoricalArray{S,1,R}(rows)
+allocate{T}(::Type{T}, rows) = Vector{T}(rows)
+allocate{T}(::Type{Vector{T}}, rows) = Vector{T}(rows)
 
 function DataTable{T <: Data.StreamType}(sch::Data.Schema,
-                                         ::Type{T}=Data.Field,
-                                         append::Bool=false,
-                                         ref::Vector{UInt8}=UInt8[], args...)
+                                         ::Type{T}=Data.Row,
+                                         append::Bool=false)
     rows, cols = size(sch)
-    rows = max(0, T <: Data.Column ? 0 : rows) # don't pre-allocate for Column streaming
+    rows = max(0, T <: Data.Batch ? 0 : rows) # don't pre-allocate for Column streaming
     columns = Vector{Any}(cols)
     types = Data.types(sch)
     for i = 1:cols
-        columns[i] = allocate(types[i], rows, ref)
+        columns[i] = allocate(types[i], rows)
     end
     return DataTable(columns, map(Symbol, Data.header(sch)))
 end
 
 # given an existing DataTable (`sink`), make any necessary changes for streaming source
 # with Data.Schema `sch` to it, given we know if we'll be `appending` or not
-function DataTable(sink, sch::Data.Schema, ::Type{Data.Field}, append::Bool,
-                   ref::Vector{UInt8})
+function DataTable(sink, sch::Data.Schema, ::Type{Data.Row}, append::Bool)
     rows, cols = size(sch)
     newsize = max(0, rows) + (append ? size(sink, 1) : 0)
-    # need to make sure we don't break a NullableVector{WeakRefString{UInt8}} when appending
```

Comment:
I'm considering a different approach where WeakRefStrings would allow you to register memory to keep track of the underlying memory. Haven't worked out the details yet, but it's next on my list.